File: 146-lru-cache/146-lru-cache.py
class DoubleLinkedList:
    def __init__(self,key:int,val:int):
        self.key = key
        self.val = val
        self.next = None
        self.pre = None


class LRUCache:

    def __init__(self, capacity: int):
        self.head = DoubleLinkedList(0,0)
        self.tail = DoubleLinkedList(0,0)
        self.head.next = self.tail
        self.tail.pre = self.head
        self.cache = {}
        self.capacity = capacity
        self.length = 0

    def get(self, key: int) -> int:
        # find node and move it to the tail
        if key not in self.cache:
            return -1
        node = self.cache[key] # get the node
        self.moveToTail(node)
        # print(self.cache.keys())
        # check = self.head
        # while check != None:
        #     print(check.key,end = " ")
        #     check = check.next
        # print("")
        return node.val
        

    def put(self, key: int, value: int) -> None:
        node = DoubleLinkedList(key,value)
        if key not in self.cache:
            self.cache[key] = node
            self.addLast(node)
            self.length += 1
            if self.length > self.capacity:
                deletedNode = self.deleteHead()
                del self.cache[deletedNode.key]
                self.length -= 1
        else:
            oldNode = self.cache[key]
            self.deleteNode(oldNode)
            self.cache[key] = node
            self.addLast(node)
    
    def moveToTail(self,node):
        self.deleteNode(node)
        self.addLast(node)
    def deleteNode(self,node):
        node.pre.next = node.next
        node.next.pre = node.pre
    def addLast(self,node):
        node.next = self.tail
        node.pre = self.tail.pre
        self.tail.pre.next = node
        self.tail.pre = node
    def deleteHead(self):
        node = self.head.next
        self.deleteNode(node)
        return node
        
    

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/146-lru-cache.java
class Node{
    int key;
    int val;
    Node next;
    Node pre;
    public Node(int key , int val){
        this.key = key;
        this.val = val;
    }
}

class LRUCache {
    int capacity;
    Node head = new Node(-1,-1);
    Node tail = new Node(-1,-1);
    HashMap<Integer,Node> cache = new HashMap<>();
    public LRUCache(int capacity) {
        head.next = tail; tail.pre = head;
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if (!cache.containsKey(key))
            return -1;
        Node node = cache.get(key);
        moveToHead(node);
        return node.val;
    }
    
    public void put(int key, int value) {
        if (!cache.containsKey(key)) {
            Node node = new Node(key, value);
            addHead(node);
            cache.put(key,node);
            capacity --;
            if (capacity == -1){
                Node last = delTail();
                cache.remove(last.key);
                capacity ++;
            }
        } else {
            Node node = cache.get(key);
            node.val = value;
            moveToHead(node);
        }
        return ;
    }
    public void addHead(Node node){
        Node next = head.next;
        head.next = node; node.next = next;
        next.pre = node ; node.pre = head;
        return ;
    }
    public void del(Node node){
        Node pre = node.pre;
        Node next = node.next;
        pre.next = next; 
        next.pre = pre;
        return ;
    }
    public Node delTail(){
        Node last = tail.pre;
        del(last);
        return last ;
    }
    public void moveToHead(Node node){
        del(node);
        addHead(node);
        return ;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */